### PR TITLE
Bug fix: missing attention mask in VAE encoder in ACT policy

### PR DESCRIPTION
**Pull request by thomwolf**
## What this does
This pull request fix the missing masking in the VAE encoder in ACT modeling.
When we train, we sample various length of episodes so we have a mask and padding tokens when a sample is less long than the sequence of action to predict. We need to handle this padding for correct VAE encoding during training.

## How it was tested
Tested the changes by adding an additional `test_backward_compatibility` configuration where the number of predicted action steps is longer than the length of the episodes in the dataset, hence padding is used for every sample.

- Added a parametrized test `("aloha", "act_1000_actions", []),` in `tests/test_policies.py` and an assiciated configuration.

Note that there could be other way to test this which would not involve adding a new configuation env in the lib, in particular:
- editing the saving format for the backward policy saved safetensors so that they include the overriden arguments in the naming and we could then use an override argument instead of a new policy (currently doing that overide the previous aloha/act test
- editing the previous aloha/act test to test the attention masking itself ( having both tested in the same run) by using the 1000 predicted steps instead of the 10 step previously tested.

I'm quite agnostic to the way we decide to organize the tests so happy to change one way or another.

## How to checkout & try? (for the reviewer)

Examples:
```bash
CUDA_VISIBLE_DEVICES="" pytest ./tests/test_policies.py::test_backward_compatibility\[aloha-act-extra_overrides2\]
```

- - -
This change is [Reviewable](https://reviewable.io/reviews/huggingface/lerobot/279)

### CODE REVIEW

**Review by alexander-soare — CHANGES_REQUESTED** [review]
Thanks a lot for this @thomwolf. I've been waiting for this one as I want to test something out.

Regarding the test, do you know if the huge chunk size slows things down significantly? If so, maybe we can also make the model a lot smaller. See my comment about passing overrides via Python rather than adding a whole new config. Then we'd need to do something like `["policy.dim_model=64", "policy.dim_feedforward=64", "policy.n_encoder_layers=2", "policy.n_vae_encoder_layers=2"]`.

File: lerobot/common/policies/act/modeling_act.py [hunk 1]
```diff
@@ -314,9 +314,16 @@ def forward(self, batch: dict[str, Tensor]) -> tuple[Tensor, tuple[Tensor, Tenso
             # Note: detach() shouldn't be necessary but leaving it the same as the original code just in case.
             pos_embed = self.vae_encoder_pos_enc.clone().detach()  # (1, S+2, D)
 
+            cls_joint_is_pad = torch.full((batch_size, 2), False).to(
+                batch["observation.state"].device
+            )  # False: not a padding
+            key_padding_mask = torch.cat([cls_joint_is_pad, batch["action_is_pad"]], axis=1)  # (bs, seq+1)
+
```
[the remark this answers is not in the document]
Comment (thomwolf) [reply lerobot/common/policies/act/modeling_act.py:321]: Thanks actually fixed my fix because the padding will be 1 or 2 depending whether we use input state or not. Added comments

File: tests/scripts/save_policy_to_safetensors.py [hunk 2]
```diff
@@ -108,15 +108,15 @@ def save_policy_to_safetensors(output_dir, env_name, policy_name, extra_override
...
-        ("aloha", "act", ["policy.n_action_steps=10"]),
-        ("dora_aloha_real", "act_real", ["policy.n_action_steps=10"]),
-        ("dora_aloha_real", "act_real_no_state", ["policy.n_action_steps=10"]),
+        # ("xarm", "tdmpc", []),
+        # (
+        #     "pusht",
+        #     "diffusion",
+        #     ["policy.n_action_steps=8", "policy.num_inference_steps=10", "policy.down_dims=[128, 256, 512]"],
+        # ),
+        ("aloha", "act_1000_actions", []),
```
[the remark this answers is not in the document]
Comment (thomwolf) [reply tests/scripts/save_policy_to_safetensors.py:117]: Maybe @aliberts you have an opinion on this question around tests and config file.

See my `How it was tested` section in this PR description.

[the remark this answers is not in the document]
Comment (alexander-soare) [reply tests/scripts/save_policy_to_safetensors.py:117]: Oh sorry, I meant also to change to using just the act.json file. Also happy to hear what @aliberts says.

**Review by alexander-soare — CHANGES_REQUESTED** [review]

File: tests/scripts/save_policy_to_safetensors.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (thomwolf) [reply tests/scripts/save_policy_to_safetensors.py:117]: oh yes this also doesn't work because the safetensor files overide themselves (cf my comment in the PR description. maybe a bit short) unless we either:
- decide we test always in this 1000 action setup for all policy tests
- decide to update the way we name safetensor files

**Comment by thomwolf** [comment]
Ok updated so that test fixtures are clean now.

There is one failing test but it's independent of this PR and the same which is also failing on main (`tests/test_policies.py::test_backward_compatibility[xarm-tdmpc-extra_overrides0]`)

Should be ready to merge whenever you think it make sense

File: tests/scripts/save_policy_to_safetensors.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aliberts) [reply tests/scripts/save_policy_to_safetensors.py]: > unless we either:
> * decide we test always in this 1000 action setup for the other ACT policy tests as well
> * decide to update the way we name safetensor files

IMO both are fine (the naming trick you went with is okay, although it adds artifacts they're relatively small). I was concerned with speed too but from what I'm seeing in CI tests it doesn't increase tests duration (cf [this branch](https://github.com/huggingface/lerobot/actions/runs/9578846846/job/26409975290?pr=279#step:7:258) vs [main](https://github.com/huggingface/lerobot/actions/runs/9565050029/job/26367137322#step:7:255))

Long-term though, I think it'll be much better to have more fine-grained tests and tests individual components of the policies and do away with these artifacts, similar to what's done in transformers. My motivations for this are:
- Whenever this test breaks, it doesn't tell us *which* part broke it.
- Having this many tests artifacts as we do right now is really not great for cloning the repo.

For now it's okay because this granularity allows us to iterate faster while still having guardrails. Happy to hear your opinions on this as well.

**Review by alexander-soare — APPROVED** [review]
